tdrs-backend/tdpservice/search_indexes/parsers/preparser.py

```diff
@@ -0,0 +1,209 @@
+    }
+
+    is_valid = validator.validate(trailer)
+
+    logger.debug("Trailer title: '%s'", trailer['title'])
+    logger.debug("Trailer record count: '%s'", trailer['record_count'])
+    logger.debug("Trailer blank: '%s'", trailer['blank'])
+    logger.debug("Trailer errors: '%s'", validator.errors)
+
+    return is_valid, validator
+
+def get_header_line(datafile):
+    """Alters header line into string."""
+    # intentionally only reading first line of file
+    datafile.seek(0)
+    line = datafile.readline()
+    datafile.seek(0)  # reset file pointer to beginning of file
+
+    # datafile when passed via redis/db is a FileField which returns bytes
+    if isinstance(line, bytes):
+        line = line.decode()
+
+    if get_record_type(line) != 'HE':
+        raise ValueError('First line in file is not recognized as a valid header.')
+    elif len(line) != 24:
+        logger.debug("line: '%s' len: %d", line, len(line))
+        return False, {'preparsing': 'Header length incorrect.'}
+    line = line.strip('\n')
+
+    return True, line
+
+def get_trailer_line(datafile):
+    """Alters the trailer line into usable string."""
+    # certify/transform input line to be correct form/type
+
+    # Don't want to read whole file, just last line, only possible with binary
+    # Credit: https://openwritings.net/pg/python/python-read-last-line-file
+    # https://stackoverflow.com/questions/46258499/
+    try:
+        datafile.seek(-2, os.SEEK_END)  # Jump to the second last byte.
+        while datafile.read(1) != b'\n':  # Check if new line.
+            datafile.seek(-2, os.SEEK_CUR)   # Jump two bytes back
+    except OSError:  # Either file is empty or contains one line.
+        datafile.seek(0)
+        return False, {'preparsing': 'File too short or missing trailer.'}
+
+    # Having set the file pointer to the last line, read it in.
+    line = datafile.readline().decode()
+    datafile.seek(0)  # Reset file pointer to beginning of file.
+
+    logger.info("Trailer line: '%s'", line)
+    if get_record_type(line) != 'TR':
+        raise ValueError('Last line is not recognized as a trailer line.')
+    elif len(line) != 24:
+        logger.debug("line: '%s' len: %d", line, len(line))
+        return False, {'preparsing': 'Trailer length incorrect.'}
+    line = line.strip('\n')
+
+    return True, line
+
+def preparse(data_file, data_type, section):
+    """Validate metadata then dispatches file to appropriate parser."""
+    if isinstance(data_file, DataFile):
+        logger.debug("Beginning preparsing on '%s'", data_file.file.name)
+        datafile = data_file.file  # do I need to open() this?
+    elif isinstance(data_file, BufferedReader):
+        datafile = data_file
+    else:
+        logger.error("Unexpected datafile type %s", type(data_file))
+        raise TypeError("Unexpected datafile type.")
+
+    header_preparsed, line = get_header_line(datafile)
+    if header_preparsed is False:
+        return False, line
```

Comment:
TODO: change this and the trailer return to a ValueError exception so the logs don't silently fail. The scheduling.parser_task doesn't inspect the return value for its preparse call.